### ISSUE TITLE
model state instantiation api

### ISSUE THREAD

**Issue by githubpsyche** [githubpsyche/jaxcmr#4]
Having functions that take functions as arguments and returns a new function seems to be a foundational idea in functional programming. Scipy's differential_evolution fn that I use for much of my model fitting takes a function as an argument and uses it to perform fitting. For the higher-order function to work, the argument function has to follow an interface documented by the higher-order function. 

In differential_evolution's case, that interface is documented as following:

> The objective function to be minimized. Must be in the form f(x, *args), where x is the argument in the form of a 1-D array and args is a tuple of any additional fixed parameters needed to completely specify the function. The number of parameters, N, is equal to len(x).

A lot of lax functions such as `jax.lax.map` are similarly good examples of higher-order functions and they also require argument functions to follow a particular interface.

**Comment by githubpsyche** [comment]
Input-Output Design Pattern

Try to write any function by answering these two questions first:

What is this function’s input (what arguments it accepts)?
What is this function’s output (what data it returns)?

Answers to these questions provide test specifications along with constraining implementation details. 

You may further use abstract types for your answers. For example, a function may accept a list of apples and return a happy fox. Such type abstractions detach the call signature from the implementation even further.

**Comment by githubpsyche** [comment]
Currying vs Partial Application

For a function of the form `f(x, *args)`, currying produces a function that takes a literal of `x` (`lx`) as an argument and returns another function of the form `f(*args)` that applies `f(lx, *args)`. 

An example from javascript:

```javascript
function curry(f) { // curry(f) does the currying transform
  return function(a) {
    return function(b) {
      return f(a, b);
    };
  };
}

// usage
function sum(a, b) {
  return a + b;
}

let curriedSum = curry(sum);

alert( curriedSum(1)(2) ); // 3
```

We can think of curried functions as factory functions for partially applied functions. `curriedSum(1)` is a partial function that fixes `a` to 1. Currying lets us easily generate partial functions.

**Comment by githubpsyche** [comment]
So what am I seeking here?

The data likelihood functions that I use to fit models usually depend on the model, the relevant data, and the parameter configuration for the model. Loss functions for fitting methods should usually consider a parameter configuration. This suggests I might benefit from partial functions that fix the model and data used across calls. Since I usually fit by subject, I expect to need many partial functions where the fixed data parameter values (while the model class parameter is held constant). This suggests that I might benefit from some currying, sure. 

Why do things this way? It hopefully means more code reuse. I won't have to write a factory function for curried functions or the like; instead, I can call my curry helper function, and the factory will be instantiated.

**Comment by githubpsyche** [comment]
Where do I start?

First, I should ensure for every concrete memorysearch class that a create_fn taking item_count, presentation_count, and parameters dict as its arguments. 

This seems to already be achieved. My fitting functions will be higher order functions that take as arguments a create_fn that accepts these arguments and returns a memorysearch instance.
